Add downgrade suggestion if database version too high

We'll be upgrading the database to v15 (likely in AnkiDroid 2.14 or 2.15), we should have a warning if the database version is too high to open with the current version of AnkiDroid

Current error:

```
2020-09-13 01:59:09.474 6493-6493/com.ichi2.anki E/Storage: Error opening collection; closing database
    com.ichi2.utils.JSONException: org.json.JSONException: A JSONObject text must begin with '{' at character 0 of 
        at com.ichi2.utils.JSONObject.<init>(JSONObject.java:129)
        at com.ichi2.utils.JSONObject.<init>(JSONObject.java:134)
        at com.ichi2.libanki.Collection.load(Collection.java:276)
        at com.ichi2.libanki.Collection.<init>(Collection.java:187)
        at com.ichi2.libanki.Storage.Collection(Storage.java:68)
        at com.ichi2.anki.CollectionHelper.getCol(CollectionHelper.java:122)
        at com.ichi2.anki.CollectionHelper.getCol(CollectionHelper.java:106)
        at com.ichi2.anki.CollectionHelper.getColSafe(CollectionHelper.java:145)
        at com.ichi2.anki.DeckPicker.onPrepareOptionsMenu(DeckPicker.java:630)
        at android.app.Activity.onPreparePanel(Activity.java:3698)
        at androidx.fragment.app.FragmentActivity.onPrepareOptionsPanel(FragmentActivity.java:501)
        at androidx.fragment.app.FragmentActivity.onPreparePanel(FragmentActivity.java:487)
        at androidx.appcompat.view.WindowCallbackWrapper.onPreparePanel(WindowCallbackWrapper.java:99)
        at androidx.appcompat.app.AppCompatDelegateImpl$AppCompatWindowCallback.onPreparePanel(AppCompatDelegateImpl.java:3097)
        at androidx.appcompat.view.WindowCallbackWrapper.onPreparePanel(WindowCallbackWrapper.java:99)
        at androidx.appcompat.app.ToolbarActionBar$ToolbarCallbackWrapper.onPreparePanel(ToolbarActionBar.java:523)
        at androidx.appcompat.app.ToolbarActionBar.populateOptionsMenu(ToolbarActionBar.java:457)
        at androidx.appcompat.app.ToolbarActionBar$1.run(ToolbarActionBar.java:57)
        at android.os.Handler.handleCallback(Handler.java:907)
        at android.os.Handler.dispatchMessage(Handler.java:105)
        at android.os.Looper.loop(Looper.java:216)
        at android.app.ActivityThread.main(ActivityThread.java:7625)
        at java.lang.reflect.Method.invoke(Native Method)
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:524)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:987)
     Caused by: org.json.JSONException: A JSONObject text must begin with '{' at character 0 of 
        at org.json.JSONTokener.syntaxError(JSONTokener.java:449)
        at com.ichi2.utils.JSONObject.<init>(JSONObject.java:76)
        at com.ichi2.utils.JSONObject.<init>(JSONObject.java:134) 
        at com.ichi2.libanki.Collection.load(Collection.java:276) 
        at com.ichi2.libanki.Collection.<init>(Collection.java:187) 
        at com.ichi2.libanki.Storage.Collection(Storage.java:68) 
        at com.ichi2.anki.CollectionHelper.getCol(CollectionHelper.java:122) 
        at com.ichi2.anki.CollectionHelper.getCol(CollectionHelper.java:106) 
        at com.ichi2.anki.CollectionHelper.getColSafe(CollectionHelper.java:145) 
        at com.ichi2.anki.DeckPicker.onPrepareOptionsMenu(DeckPicker.java:630) 
        at android.app.Activity.onPreparePanel(Activity.java:3698) 
        at androidx.fragment.app.FragmentActivity.onPrepareOptionsPanel(FragmentActivity.java:501) 
        at androidx.fragment.app.FragmentActivity.onPreparePanel(FragmentActivity.java:487) 
        at androidx.appcompat.view.WindowCallbackWrapper.onPreparePanel(WindowCallbackWrapper.java:99) 
        at androidx.appcompat.app.AppCompatDelegateImpl$AppCompatWindowCallback.onPreparePanel(AppCompatDelegateImpl.java:3097) 
        at androidx.appcompat.view.WindowCallbackWrapper.onPreparePanel(WindowCallbackWrapper.java:99) 
        at androidx.appcompat.app.ToolbarActionBar$ToolbarCallbackWrapper.onPreparePanel(ToolbarActionBar.java:523) 
        at androidx.appcompat.app.ToolbarActionBar.populateOptionsMenu(ToolbarActionBar.java:457) 
        at androidx.appcompat.app.ToolbarActionBar$1.run(ToolbarActionBar.java:57) 
        at android.os.Handler.handleCallback(Handler.java:907) 
        at android.os.Handler.dispatchMessage(Handler.java:105) 
        at android.os.Looper.loop(Looper.java:216) 
        at android.app.ActivityThread.main(ActivityThread.java:7625) 
        at java.lang.reflect.Method.invoke(Native Method) 
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:524) 
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:987) 
```

--- a/AnkiDroid/src/main/java/com/ichi2/anki/CollectionHelper.java
+++ b/AnkiDroid/src/main/java/com/ichi2/anki/CollectionHelper.java
@@ -28,6 +28,7 @@ import androidx.annotation.Nullable;
 import com.ichi2.anki.exception.StorageAccessException;
 import com.ichi2.libanki.Collection;
 import com.ichi2.libanki.Storage;
+import com.ichi2.libanki.exception.UnknownDatabaseVersionException;
 import com.ichi2.libanki.utils.SystemTime;
 import com.ichi2.libanki.utils.Time;
 import com.ichi2.preferences.PreferenceExtensions;
@@ -38,6 +39,8 @@ import java.io.IOException;
 
 import androidx.annotation.VisibleForTesting;
 import timber.log.Timber;
+
+import static com.ichi2.libanki.Consts.SCHEMA_VERSION;
 
 /**
  * Singleton which opens, stores, and closes the reference to the Collection.
@@ -371,6 +374,22 @@ public class CollectionHelper {
      */
     public static void loadCollectionComplete(Collection col) {
         col.getModels();
+    }
+
+    public static boolean isFutureAnkiDroidVersion(Context context) throws UnknownDatabaseVersionException {
+        int databaseVersion = getDatabaseVersion(context);
+        return databaseVersion > SCHEMA_VERSION;
+    }
+
+
+    public static int getDatabaseVersion(Context context) throws UnknownDatabaseVersionException {
+        try {
+            Collection col = getInstance().mCollection;
+            return col.queryVer();
+        } catch (Exception e) {
+            Timber.w(e, "Failed to query version");
+            return Storage.getDatabaseVersion(getCollectionPath(context));
+        }
     }
 
 }

--- a/AnkiDroid/src/main/java/com/ichi2/anki/DeckPicker.java
+++ b/AnkiDroid/src/main/java/com/ichi2/anki/DeckPicker.java
@@ -520,6 +520,9 @@ public class DeckPicker extends NavigationDrawerActivity implements
                     Intent i = Preferences.getPreferenceSubscreenIntent(this, "com.ichi2.anki.prefs.advanced");
                     startActivityForResultWithoutAnimation(i, REQUEST_PATH_UPDATE);
                     Toast.makeText(this, R.string.directory_inaccessible, Toast.LENGTH_LONG).show();
+                } else if (isFutureAnkiDroidVersion()) {
+                    Timber.i("Displaying database versioning");
+                    showDatabaseErrorDialog(DatabaseErrorDialog.INCOMPATIBLE_DB_VERSION);
                 } else {
                     Timber.i("Displaying database error");
                     showDatabaseErrorDialog(DatabaseErrorDialog.DIALOG_LOAD_FAILED);
@@ -529,6 +532,17 @@ public class DeckPicker extends NavigationDrawerActivity implements
 
         mShortAnimDuration = getResources().getInteger(android.R.integer.config_shortAnimTime);
     }
+
+
+    private boolean isFutureAnkiDroidVersion() {
+        try {
+            return CollectionHelper.isFutureAnkiDroidVersion(this);
+        } catch (Exception e) {
+            Timber.w(e, "Could not determine if future AnkiDroid version - assuming not");
+            return false;
+        }
+    }
+
 
     // throws doesn't seem to be checked by the compiler - consider it to be documentation
     private boolean applyDeckPickerBackground(View view) throws OutOfMemoryError {

--- a/AnkiDroid/src/main/java/com/ichi2/anki/dialogs/DatabaseErrorDialog.java
+++ b/AnkiDroid/src/main/java/com/ichi2/anki/dialogs/DatabaseErrorDialog.java
@@ -13,12 +13,15 @@ import com.ichi2.anki.BackupManager;
 import com.ichi2.anki.CollectionHelper;
 import com.ichi2.anki.DeckPicker;
 import com.ichi2.anki.R;
+import com.ichi2.libanki.Consts;
 import com.ichi2.async.Connection;
 import com.ichi2.libanki.utils.Time;
+import com.ichi2.utils.UiUtil;
 
 import java.io.File;
 import java.io.IOException;
 import java.util.ArrayList;
+import java.util.List;
 
 import androidx.annotation.NonNull;
 import timber.log.Timber;
@@ -38,6 +41,8 @@ public class DatabaseErrorDialog extends AsyncDialogFragment {
     public static final int DIALOG_FULL_SYNC_FROM_SERVER = 8;
     /** If the database is locked, all we can do is reset the app */
     public static final int DIALOG_DB_LOCKED = 9;
+    /** If the datbase is at a version higher than what we can currently handle */
+    public static final int INCOMPATIBLE_DB_VERSION = 10;
 
     // public flag which lets us distinguish between inaccessible and corrupt database
     public static boolean databaseCorruptFlag = false;
@@ -209,7 +214,6 @@ public class DatabaseErrorDialog extends AsyncDialogFragment {
                     }
                     builder.title(res.getString(R.string.backup_restore_select_title))
                             .negativeText(res.getString(R.string.dialog_cancel))
-                            .onNegative((inner_dialog, which) -> dismissAllDialogFragments())
                             .items(dates)
                             .itemsCallbackSingleChoice(dates.length,
                                     (materialDialog, view, which, charSequence) -> {
@@ -298,6 +302,31 @@ public class DatabaseErrorDialog extends AsyncDialogFragment {
                         .onPositive((inner_dialog, which) -> exit())
                         .show();
             }
+            case INCOMPATIBLE_DB_VERSION: {
+                List<Integer> values = new ArrayList<>();
+                CharSequence[] options = new CharSequence[] { UiUtil.makeBold(res.getString(R.string.backup_restore)), UiUtil.makeBold(res.getString(R.string.backup_full_sync_from_server)) };
+                values.add(0);
+                values.add(1);
+                return builder
+                        .cancelable(false)
+                        .content(getMessage())
+                        .iconAttr(R.attr.dialogErrorIcon)
+                        .positiveText(res.getString(R.string.close))
+                        .onPositive((inner_dialog, which) -> exit())
+                        .items(options)
+                       // .itemsColor(ContextCompat.getColor(requireContext(), R.color.material_grey_500))
+                        .itemsCallback((dialog, itemView, position, text) -> {
+                            switch (values.get(position)) {
+                                case 0:
+                                    ((DeckPicker) getActivity()).showDatabaseErrorDialog(DIALOG_RESTORE_BACKUP);
+                                    break;
+                                case 1:
+                                    ((DeckPicker) getActivity()).showDatabaseErrorDialog(DIALOG_FULL_SYNC_FROM_SERVER);
+                                    break;
+                            }
+                        })
+                        .show();
+            }
             default:
                 return null;
         }
@@ -337,6 +366,14 @@ public class DatabaseErrorDialog extends AsyncDialogFragment {
                 return res().getString(R.string.backup_full_sync_from_server_question);
             case DIALOG_DB_LOCKED:
                 return res().getString(R.string.database_locked_summary);
+            case INCOMPATIBLE_DB_VERSION:
+                int databaseVersion = -1;
+                try {
+                    databaseVersion = CollectionHelper.getDatabaseVersion(requireContext());
+                } catch (Exception e) {
+                    Timber.w(e, "Failed to get database version, using -1");
+                }
+                return res().getString(R.string.incompatible_database_version_summary, Consts.SCHEMA_VERSION, databaseVersion);
             default:
                 return getArguments().getString("dialogMessage");
         }
@@ -362,6 +399,8 @@ public class DatabaseErrorDialog extends AsyncDialogFragment {
                 return res().getString(R.string.backup_full_sync_from_server);
             case DIALOG_DB_LOCKED:
                 return res().getString(R.string.database_locked_title);
+            case INCOMPATIBLE_DB_VERSION:
+                return res().getString(R.string.incompatible_database_version_title);
             case DIALOG_DB_ERROR:
             default:
                 return res().getString(R.string.answering_error_title);

--- a/AnkiDroid/src/main/java/com/ichi2/libanki/Collection.java
+++ b/AnkiDroid/src/main/java/com/ichi2/libanki/Collection.java
@@ -33,6 +33,7 @@ import com.ichi2.anki.analytics.UsageAnalytics;
 import com.ichi2.anki.exception.ConfirmModSchemaException;
 import com.ichi2.async.CollectionTask;
 import com.ichi2.libanki.exception.NoSuchDeckException;
+import com.ichi2.libanki.exception.UnknownDatabaseVersionException;
 import com.ichi2.libanki.hooks.ChessFilter;
 import com.ichi2.libanki.sched.AbstractSched;
 import com.ichi2.libanki.sched.Sched;
@@ -2138,6 +2139,14 @@ public class Collection {
     @CheckResult
     public List<Long> filterToValidCards(long[] cards) {
         return getDb().queryLongList("select id from cards where id in " + Utils.ids2str(cards));
+    }
+
+    public int queryVer() throws UnknownDatabaseVersionException {
+        try {
+            return getDb().queryScalar("select ver from col");
+        } catch (Exception e) {
+            throw new UnknownDatabaseVersionException(e);
+        }
     }
 
     //This duplicates _loadScheduler (but returns the value and sets the report limit).

--- a/AnkiDroid/src/main/java/com/ichi2/libanki/Storage.java
+++ b/AnkiDroid/src/main/java/com/ichi2/libanki/Storage.java
@@ -21,6 +21,7 @@ import android.content.Context;
 
 import com.ichi2.anki.exception.ConfirmModSchemaException;
 
+import com.ichi2.libanki.exception.UnknownDatabaseVersionException;
 import com.ichi2.libanki.utils.SystemTime;
 import com.ichi2.libanki.utils.Time;
 import com.ichi2.utils.JSONArray;
@@ -44,6 +45,16 @@ public class Storage {
         return Collection(context, path, false, false);
     }
 
+    /** Helper method for when the collection can't be opened */
+    public static int getDatabaseVersion(String path) throws UnknownDatabaseVersionException {
+        try {
+            DB db = new DB(path);
+            return db.queryScalar("SELECT ver FROM col");
+        } catch (Exception e) {
+            Timber.w(e, "Can't open database");
+            throw new UnknownDatabaseVersionException(e);
+        }
+    }
 
     public static Collection Collection(Context context, String path, boolean server, boolean log) {
         return Collection(context, path, server, log, new SystemTime());

--- a/AnkiDroid/src/main/java/com/ichi2/libanki/exception/UnknownDatabaseVersionException.java
+++ b/AnkiDroid/src/main/java/com/ichi2/libanki/exception/UnknownDatabaseVersionException.java
@@ -1,0 +1,23 @@
+/*
+ *  Copyright (c) 2020 David Allison <davidallisongithub@gmail.com>
+ *
+ *  This program is free software; you can redistribute it and/or modify it under
+ *  the terms of the GNU General Public License as published by the Free Software
+ *  Foundation; either version 3 of the License, or (at your option) any later
+ *  version.
+ *
+ *  This program is distributed in the hope that it will be useful, but WITHOUT ANY
+ *  WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
+ *  PARTICULAR PURPOSE. See the GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License along with
+ *  this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package com.ichi2.libanki.exception;
+
+public class UnknownDatabaseVersionException extends Exception {
+    public UnknownDatabaseVersionException(Exception exception) {
+        super(exception);
+    }
+}

--- a/AnkiDroid/src/main/java/com/ichi2/utils/UiUtil.java
+++ b/AnkiDroid/src/main/java/com/ichi2/utils/UiUtil.java
@@ -1,0 +1,29 @@
+/*
+ *  Copyright (c) 2020 David Allison <davidallisongithub@gmail.com>
+ *
+ *  This program is free software; you can redistribute it and/or modify it under
+ *  the terms of the GNU General Public License as published by the Free Software
+ *  Foundation; either version 3 of the License, or (at your option) any later
+ *  version.
+ *
+ *  This program is distributed in the hope that it will be useful, but WITHOUT ANY
+ *  WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
+ *  PARTICULAR PURPOSE. See the GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License along with
+ *  this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package com.ichi2.utils;
+
+import android.text.Spannable;
+import android.text.SpannableStringBuilder;
+import android.text.style.StyleSpan;
+
+public class UiUtil {
+    public static Spannable makeBold(String s) {
+        SpannableStringBuilder str = new SpannableStringBuilder(s);
+        str.setSpan(new StyleSpan(android.graphics.Typeface.BOLD), 0, s.length(), Spannable.SPAN_EXCLUSIVE_EXCLUSIVE);
+        return str;
+    }
+}

--- a/AnkiDroid/src/main/res/values/03-dialogs.xml
+++ b/AnkiDroid/src/main/res/values/03-dialogs.xml
@@ -191,6 +191,10 @@
     <string name="database_locked_title">Database Locked</string>
     <string name="database_locked_summary">The AnkiDroid database is in use by another application. Please close the other application then reopen AnkiDroid.</string>
 
+
+    <string name="incompatible_database_version_title">Incompatible Database Version</string>
+    <string name="incompatible_database_version_summary">The database is a more advanced version than this version of AnkiDroid can work with. Upgrade AnkiDroid or downgrade the database to open it\n\nSupported version: %1$d\nDatabase version: %2$d\n\nThe following restore options will overwrite your current collection, possibly with a compatible database version:</string>
+
     <!-- Deck Picker Background-->
     <string name="background_image_applied">Background image applied</string>
     <string name="background_image_removed">Background image removed</string>

